test/extensions/io_socket/buffered_io_socket/user_space_file_event_impl_test.cc

```diff
@@ -0,0 +1,240 @@
+    SCOPED_TRACE(absl::StrCat("current event:", current_event));
+    EXPECT_CALL(io_source_, isReadable()).WillOnce(Return(true)).RetiresOnSaturation();
+    EXPECT_CALL(io_source_, isPeerWritable()).WillOnce(Return(true)).RetiresOnSaturation();
+    user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+        *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+    // user_file_event_->activate(e);
+    EXPECT_CALL(ready_cb_, called(current_event));
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+    user_file_event_.reset();
+  }
+}
+
+TEST_F(UserSpaceFileEventImplTest, TestActivateWillSchedule) {
+  // IO is neither readable nor writable.
+  user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+      *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+  {
+    EXPECT_CALL(ready_cb_, called(_)).Times(0);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+  {
+    user_file_event_->activate(Event::FileReadyType::Read);
+    EXPECT_CALL(ready_cb_, called(Event::FileReadyType::Read)).Times(1);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+  {
+    user_file_event_->activate(Event::FileReadyType::Write);
+    EXPECT_CALL(ready_cb_, called(Event::FileReadyType::Write)).Times(1);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+}
+
+TEST_F(UserSpaceFileEventImplTest, TestActivateDedup) {
+  // IO is neither readable nor writable.
+  user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+      *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+  {
+    EXPECT_CALL(ready_cb_, called(_)).Times(0);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+  {
+    user_file_event_->activate(Event::FileReadyType::Read);
+    user_file_event_->activate(Event::FileReadyType::Write);
+    user_file_event_->activate(Event::FileReadyType::Write);
+    user_file_event_->activate(Event::FileReadyType::Read);
+    EXPECT_CALL(ready_cb_, called(event_rw)).Times(1);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+  {
+    EXPECT_CALL(ready_cb_, called(_)).Times(0);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+}
+
+TEST_F(UserSpaceFileEventImplTest, TestEnabledClearActivate) {
+  // IO is neither readable nor writable.
+  user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+      *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+  {
+    EXPECT_CALL(ready_cb_, called(_)).Times(0);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  {
+    setWritable();
+    setReadable();
+    user_file_event_->activate(Event::FileReadyType::Read);
+    user_file_event_->setEnabled(Event::FileReadyType::Write);
+    EXPECT_CALL(ready_cb_, called(Event::FileReadyType::Write)).Times(1);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
```

Comment:
Added

Reply:
Not sure what I added in the past... But it's there now.
```
  // No event is delivered since it's either user disabled or io_handle doesn't provide pending
  // data.
  {
    clearWriteReable();
    setReadable();
    user_file_event_->activate(Event::FileReadyType::Read);
    user_file_event_->setEnabled(Event::FileReadyType::Write);
    EXPECT_CALL(ready_cb_, called(_)).Times(0);
    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
  }
  {
    clearWriteReable();
    setWritable();
    user_file_event_->activate(Event::FileReadyType::Write);
    user_file_event_->setEnabled(Event::FileReadyType::Read);
    EXPECT_CALL(ready_cb_, called(_)).Times(0);
    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
  }
```